File: packages/go/luna/resources/storage.go
package resources

import (
	"context"
	"encoding/json"
	"fmt"

	lunahttp "github.com/eclipse-softworks/luna-sdk-go/luna/http"
)

// BucketsResource provides access to bucket operations
type BucketsResource struct {
	client   *lunahttp.Client
	basePath string
}

// List retrieves all buckets
func (r *BucketsResource) List(ctx context.Context) (*BucketList, error) {
	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "GET",
		Path:   r.basePath,
	})
	if err != nil {
		return nil, err
	}

	var result BucketList
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Upload uploads a file to a bucket
// Note: This is a placeholder for the binary upload.
func (r *BucketsResource) Upload(ctx context.Context, bucketID string, body interface{}) (*FileObject, error) {
	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "POST",
		Path:   fmt.Sprintf("%s/%s/upload", r.basePath, bucketID),
		Body:   body,
	})
	if err != nil {
		return nil, err
	}

	var result FileObject
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// FilesResource provides access to file operations
type FilesResource struct {
	client   *lunahttp.Client
	basePath string
}

// GetDownloadURL retrieves the download URL for a file
func (r *FilesResource) GetDownloadURL(ctx context.Context, id string) (string, error) {
	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "GET",
		Path:   fmt.Sprintf("%s/%s/download", r.basePath, id),
	})
	if err != nil {
		return "", err
	}

	// Assuming response is {"url": "..."}
	var result struct {
		URL string `json:"url"`
	}
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return "", fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return result.URL, nil
}

// StorageResource groups Storage service resources
type StorageResource struct {
	Buckets *BucketsResource
	Files   *FilesResource
}

// NewStorageResource creates a new Storage resource
func NewStorageResource(client *lunahttp.Client) *StorageResource {
	return &StorageResource{
		Buckets: &BucketsResource{
			client:   client,
			basePath: "/v1/storage/buckets",
		},
		Files: &FilesResource{
			client:   client,
			basePath: "/v1/storage/files",
		},
	}
}

File: packages/typescript/src/resources/storage.ts
import type { HttpClient } from '../http/client.js';
import type { Bucket, BucketList, FileObject, ListResponse } from '../types/index.js';

export class BucketsResource {
    private readonly httpClient: HttpClient;
    private readonly basePath = '/v1/storage/buckets';

    constructor(httpClient: HttpClient) {
        this.httpClient = httpClient;
    }

    /**
     * List all buckets
     */
    async list(): Promise<BucketList> {
        const response = await this.httpClient.request<BucketList>({
            method: 'GET',
            path: this.basePath,
        });
        return response.data;
    }

    async get(id: string): Promise<Bucket> {
        const response = await this.httpClient.request<Bucket>({
            method: 'GET',
            path: `${this.basePath}/${id}`,
        });
        return response.data;
    }

    /**
     * Upload a file to a bucket
     * Note: In a real implementation, this might handle Multipart forms.
     * For now, we'll assume a direct binary upload or presigned URL flow wrapper.
     * Following the spec: POST /v1/storage/buckets/:id/upload
     */
    async upload(bucketId: string, file: Blob | Buffer | unknown, metadata?: Record<string, string>): Promise<FileObject> {
        // Implementation detail: constructing FormData would happen here
        // For SDK simplicity in this exercise, we pass the body directly or as structured data
        const response = await this.httpClient.request<FileObject>({
            method: 'POST',
            path: `${this.basePath}/${bucketId}/upload`,
            body: { file, metadata },
        });
        return response.data;
    }
}

export class FilesResource {
    private readonly httpClient: HttpClient;
    private readonly basePath = '/v1/storage/files';

    constructor(httpClient: HttpClient) {
        this.httpClient = httpClient;
    }

    /**
     * List files in a bucket
     */
    async list(bucketId: string): Promise<ListResponse<FileObject>> {
        const response = await this.httpClient.request<ListResponse<FileObject>>({
            method: 'GET',
            path: `/v1/storage/buckets/${bucketId}/files`,
        });
        return response.data;
    }

    /**
     * Get download URL for a file
     */
    async getDownloadUrl(id: string): Promise<{ url: string }> {
        const response = await this.httpClient.request<{ url: string }>({
            method: 'GET',
            path: `${this.basePath}/${id}/download`,
        });
        return response.data;
    }
}

export class StorageResource {
    public readonly buckets: BucketsResource;
    public readonly files: FilesResource;

    constructor(httpClient: HttpClient) {
        this.buckets = new BucketsResource(httpClient);
        this.files = new FilesResource(httpClient);
    }
}
